### ISSUE TITLE
Game Feedback

### ISSUE THREAD

**Issue by DrNeroCF** [GDquest/godot-metroidvania-2d#109]
- Hook won't look on through passthrough blocks.  With as small and the launch force is after a hook, it makes sense to put hook points on top of (or inside) passthrough blocks

- The ledge code is pretty complex considering how minimal the rest of the engine is right now, and causes some issues: - walking between two ground with a 3 tile gap in between, you'll fall pretty low but get caught and pulled up - if you're jumping at a ledge, your upward momentum will be canceled - will trigger while jumping up to passthrough blocks if you have horizontal speed

- The camera's pretty twitchy, especially with a gamepad.  I fiddled around a bit with the code: added a deadzone, made the distance relative to the joystick tilt strength

- While editing, I'd love to see a way to switch between tile map editing and object editing, maybe holding a button, or can objects still be draggable while in tile mode?

- I handed the controller to one of my kids while editing the level, but the mode switched between mouse and gamepad (otherwise placing blocks while they're still playing is really cool).  Could the mouse be limited to the game window?

- Analog movement with gamepad?  Maybe just some momentum on the player while on the ground.

- I'd like to use W to jump while using kb&m, but maybe I'm just weird like that

- I sort of wonder if we need an aimer at all?  Maybe if the hook is used as an attack later

- Hook should be shootable even when target isn't selected, if used for attack

- Might be perfectionist of me, but the player blips a frame back when landing and I think when stopping

- Pulling away from a wall should cancel the wall jump state

- Pressing R has caused the level to reset to an older version before.

- The biggest change I would like to try is boosting the momentum after hooking.  This would allow for way more placements for hooks, and I think launching yourself onto a platform would just feel really good.  Later upgrade could be something different (more boost, jump after hook, hook onto walls?).  As it is now, hooks must be placed very close to where the player is going to land, which is almost too obvious?  I'm not sure why exactly why that bugs me visually...  As it is, hooks can't even be placed under platforms.

Let me know if these should be explained in more detail, or if pictures or video would help!

**Comment by NathanLovato** [comment]
Thanks for the feedback. Everything sounds pretty clear to me.

Please leave the code to us. If you do code changes locally, please leave them out of your level design branch and related commits, or if you want review and to add it, create a dedicated git branch and open a pull request on the repo.

The code you wrote looks like flash code. There is an idiomatic style for Godot's GDScript that we want to follow, along with our own guidelines. It's important for the code style to be consistent and to show good practices as we teach programming. Also, it'll be a lot faster for us to make changes to the code as we've all been using Godot for a while and know the codebase.

**Comment by NathanLovato** [comment]
I'm writing the issues now. A few answers:

> While editing, I'd love to see a way to switch between tile map editing and object editing

There is no way to do that built-in unfortunately, you have to click on a node in the scene tree to get out of the tilemap editing mode.

> Analog movement with gamepad?

Gotta see, in the course I want us to talk about the challenges of offering a balanced experience with different controllers. While you can use a joystick to control a character's speed, you can't with the keyboard. At first the character had acceleration, I removed it some time ago because of that.

> I sort of wonder if we need an aimer at all?

It's a debug tool for now, let's see later when more is in place. You can toggle its visibility off by clicking the eye next to the node in the Player scene.

> Pressing R has caused the level to reset to an older version before.

R reloads the scene from the disk. You probably had not saved your level. Hot reload gives you a real-time preview of your changes in the game but it does not save your files. That way, if a change doesn't work, you can close the scene without saving. 

If that happens again you can focus the editor, save the scene, focus on the game window and press R again. It should work. Otherwise, it might be an engine-side bug.

> The biggest change I would like to try is boosting the momentum after hooking

We can try that. I still want to use a steering behavior to go to the hooking point as before I was using physics and the movement was too unreliable, there many corner cases.

Right now I'm using the arrive behavior, so the character slows down a little bit upon reaching the point. It gives you a few extra frames to “plan” the next hook. But we can preserve the velocity moving to the hooking point and give the character a fixed speed after reaching the target.

**Comment by NathanLovato** [comment]
I'll get to it.

Issues:

- #111
- #112
- #113
- #114
- #115
- #116
- #117
- #118

**Comment by DrNeroCF** [comment]
> Please leave the code to us. If you do code changes locally, please leave them out of your level design branch and related commits, or if you want review and to add it, create a dedicated git branch and open a pull request on the repo.

Completely understood, I kept a copy of the original files, I might be misunderstanding how separated a fork is from the original code.  I'd only use the code to help discuss if a smoother camera (not coded by me) would be better or not.  I'll make sure everything is original other than the level file.

> It's a debug tool for now, let's see later when more is in place.

I mean more about having the player aim at all.  As in, just having the node light up if it's in range, and pressing hook always goes to the closest node that's still above the player.  I'm just thinking about how much extra control is the player given versus the added complexity.  How often could there be two nodes near the player at the same time?  Just food for thought, I try to be pretty minimalistic with the required complexity for core mechanics, then build from there.  But of course, that's up to you guys in the end.

> R reloads the scene from the disk.

Ah, makes sense!

>  It gives you a few extra frames to “plan” the next hook

Right, I debating whether or not the extra speed would make chaining hooks too difficult.  But, if there's more movement there, the player would have more options for where to hook and launch too (choosing whether to launch to the left, right, or straight up), which makes the hook path less... predetermined?

**Comment by NathanLovato** [comment]
## About aiming with the stick vs auto-aim

I've tested auto-aim at first, hooking to the closest point. Even favoring hooks that are forward first, if the hook range is comfortable, I've ended going e.g. up instead of forward. 

To me, having the aim on the left stick gives you a little more air control in particular, if you start to have branching paths mid-air with hooking points, things like these. But I don't know if these controls are intuitive or hard for other people.

**Comment by DrNeroCF** [comment]
Right, definitely gives the player more control, I'm just trying to do my due diligence discussing additional difficulty vs player agency.  If there were issues with it before, it's not a problem to just run with the manual aiming.

**Comment by NathanLovato** [comment]
We've addressed or added issues for all the points above, closing this issue for now. Feel free to create a new issue to give feedback in a batch like that, or on Discord, or to create individual, specific issues if e.g. you stumble upon a new bug
